packages/react/src/components/button.tsx

```diff
@@ -0,0 +1,178 @@
+import { useRef, useState } from "react";
+import { twMerge } from "tailwind-merge";
+import { generateMimeTypes, classNames } from "uploadthing/client";
+import { INTERNAL_uploadthingHookGen } from "../useUploadThing";
+import { styleFieldToClassName, styleFieldToCssObject, spinnerFieldToElement, contentFieldToContent } from "../utils/styles";
+import { Spinner } from "./shared";
+import {
+    allowedContentTextLabelGenerator,
+    generatePermittedFileTypes,
+    progressHeights
+} from './shared'
+
+type ButtonStyleFieldCallbackArgs = {
+    ready: boolean;
+    isUploading: boolean;
+    uploadProgress: number;
+    fileTypes: string[];
+};
+
+export type UploadButtonProps<TRouter extends FileRouter> =
+    UploadthingComponentProps<TRouter> & {
+        appearance?: {
+            container?: StyleField<ButtonStyleFieldCallbackArgs>;
+            button?: StyleField<ButtonStyleFieldCallbackArgs>;
+            buttonSpinner?: SpinnerField<ButtonStyleFieldCallbackArgs>;
+            allowedContent?: StyleField<ButtonStyleFieldCallbackArgs>;
+        };
+        content?: {
+            button?: ContentField<ButtonStyleFieldCallbackArgs>;
+            allowedContent?: ContentField<ButtonStyleFieldCallbackArgs>;
+        };
+        className?: string;
+    };
+
+/**
+ * @example
+ * <UploadButton<OurFileRouter>
+ *   endpoint="someEndpoint"
+ *   onUploadComplete={(res) => console.log(res)}
+ *   onUploadError={(err) => console.log(err)}
+ * />
+ */
+export function UploadButton<TRouter extends FileRouter>(
+    props: FileRouter extends TRouter
+        ? ErrorMessage<"You forgot to pass the generic">
+        : UploadButtonProps<TRouter>,
+) {
+    // Cast back to UploadthingComponentProps<TRouter> to get the correct type
+    // since the ErrorMessage messes it up otherwise
+    const $props = props as UploadButtonProps<TRouter>;
+    const useUploadThing = INTERNAL_uploadthingHookGen<TRouter>();
+
+    const fileInputRef = useRef<HTMLInputElement>(null);
+    const [uploadProgress, setUploadProgress] = useState(0);
+    const { startUpload, isUploading, permittedFileInfo } = useUploadThing(
+        $props.endpoint,
+        {
+            onClientUploadComplete: (res) => {
+                if (fileInputRef.current) {
+                    fileInputRef.current.value = "";
+                }
+                $props.onClientUploadComplete?.(res);
+                setUploadProgress(0);
+            },
+            onUploadProgress: (p) => {
+                setUploadProgress(p);
+                $props.onUploadProgress?.(p);
+            },
+            onUploadError: $props.onUploadError,
+        },
+    );
+
+    const { fileTypes, multiple } = generatePermittedFileTypes(
+        permittedFileInfo?.config,
+    );
+
+    const ready = fileTypes.length > 0;
+
+    const getUploadButtonText = (fileTypes: string[]) => {
+        if (!(fileTypes.length > 0)) return "Loading...";
```

Comment:
gotta love this logic

Reply:
FYI, it was like that before me, but I can refactor ;D